src/Lachain.Core/Blockchain/Operations/CheckpointManager.cs

```diff
@@ -0,0 +1,234 @@
+
+namespace Lachain.Core.Blockchain.Operations
+{
+    public class CheckpointManager : ICheckpointManager
+    {
+        public static readonly uint _checkpointPeriod = 1000000; // 10^6
+        private static readonly ILogger<CheckpointManager> Logger = LoggerFactory.GetLoggerForClass<CheckpointManager>();
+        private readonly IBlockManager _blockManager;
+        private readonly ICheckpointRepository _repository;
+        private readonly ISnapshotIndexRepository _snapshotIndexer;
+        private ulong? _checkpointBlockId;
+        private UInt256? _checkpointBlockHash;
+        private IDictionary<RepositoryType, UInt256?> _stateHashes = new Dictionary<RepositoryType, UInt256?>();
+        public ulong? CheckpointBlockId => _checkpointBlockId;
+        public UInt256? CheckpointBlockHash => _checkpointBlockHash;
+        public CheckpointManager(
+            IBlockManager blockManager,
+            ICheckpointRepository repository,
+            ISnapshotIndexRepository snapshotIndexer
+        )
+        {
+            _blockManager = blockManager;
+            _repository = repository;
+            _snapshotIndexer = snapshotIndexer;
+            UpdateCache();
+            _blockManager.OnBlockPersisted += UpdateCheckpoint;
+        }
+
+        public static ulong GetNextCheckpointHeight(ulong height)
+        {
+            return _checkpointPeriod * (height / _checkpointPeriod + 1);
+        }
+
+        public static ulong GetClosestCheckpointHeight(ulong height)
+        {
+            return _checkpointPeriod * (height / _checkpointPeriod);
+        }
+        
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        private void UpdateCheckpoint(object? sender, Block block)
+        {
+            if (CheckpointReached(block.Header.Index))
+            {
+                SaveCheckpoint(block);
+            }
+        }
+
+        // Check if checkpoint reached. No need to save checkpoint for genesis block
+        private bool CheckpointReached(ulong blockId)
+        {
+            return blockId > 0 && blockId % _checkpointPeriod == 0 ;
+        }
+
+        public bool SaveCheckpoint(Block block)
+        {
+            if (_checkpointBlockId >= block.Header.Index)
+            {
+                Logger.LogWarning($"We have checkpoint block {_checkpointBlockId}, trying to save checkpoint block "
+                    + $"{block.Header.Index}. Aborting save.");
+                return false;
+            }
+            Logger.LogTrace($"Saving checkpoint for block {block.Header.Index}");
+            var saved = _repository.SaveCheckpoint(block);
+            if (saved)
+            {
+                Logger.LogTrace("Saved checkpoint successfully");
+                UpdateCache();
+            }
+            else Logger.LogTrace("Could not save checkpoint");
+            return saved;
+        }
+
+        private void UpdateCache()
+        {
+            _checkpointBlockId = _repository.FetchCheckpointBlockId();
```

Comment:
It is important security requirement: checkpoints should be hardcoded in the source code,  when somebody downloads client from the repo it gets checkpoint list,  checlpoints are created by the developer team, one should change source code to chjange checkpoints